src/datasources/asset-utilization/AssetUtilizationDataSource.ts

```diff
@@ -0,0 +1,185 @@
+import {
+    DataFrameDTO,
+    DataQueryRequest,
+    DataSourceInstanceSettings,
+    dateTime,
+} from '@grafana/data';
+import {BackendSrv, getBackendSrv, getTemplateSrv, TemplateSrv, TestingStatus} from '@grafana/runtime';
+import {DataSourceBase} from 'core/DataSourceBase';
+import {
+    AssetsResponse,
+    AssetUtilizationHistory, AssetUtilizationHistoryResponse,
+    AssetUtilizationQuery,
+    IsNIAsset,
+    IsPeak,
+    ServicePolicyModel,
+    UtilizationCategory,
+    UtilizationTimeFrequency,
+    ValidAssetUtilizationQuery,
+    Weekday
+} from './types';
+import deepEqual from "fast-deep-equal";
+import {
+    calculateDailyUtilization, convertTimeToUTC, extractTimestampesFromData,
+    filterDataByTimeRange,
+    getBusinessHours, getLocalTimezoneOffsetInHours,
+    getStartEndDates,
+    mergeOverlappingIntervals,
+    patchMissingEndTimestamps, splitIntervalsByDays
+} from "./helper";
+
+export class AssetUtilizationDataSource extends DataSourceBase<AssetUtilizationQuery> {
+    constructor(
+        readonly instanceSettings: DataSourceInstanceSettings,
+        readonly backendSrv: BackendSrv = getBackendSrv(),
+        readonly templateSrv: TemplateSrv = getTemplateSrv()
+    ) {
+        super(instanceSettings, backendSrv, templateSrv);
+    }
+
+    baseUrl = this.instanceSettings.url + '/niapm/v1';
+
+    defaultQuery: Omit<ValidAssetUtilizationQuery, 'refId'> = {
+        assetIdentifier: '',
+        isNIAsset: IsNIAsset.BOTH,
+        minionId: '',
+        isPeak: IsPeak.PEAK,
+        timeFrequency: UtilizationTimeFrequency.DAILY,
+        utilizationCategory: UtilizationCategory.ALL,
+        peakDays: [Weekday.Monday, Weekday.Tuesday, Weekday.Wednesday, Weekday.Thursday, Weekday.Friday],
+    };
+
+    async runQuery(query: AssetUtilizationQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
+        const processedQuery = this.processQuery(query);
+        const {range} = options;
+        const from = (new Date(range!.from.valueOf())).toISOString();
+        const to = (new Date(range!.to.valueOf())).toISOString();
```

Comment:
Why are we creating a new Date object both here and when we pass it to `onLoadUtilizationHistory`? Can we simplify this?

Reply:
removed